sklearn/utils/optimize.py

```diff
@@ -111,49 +177,26 @@ def newton_cg(func_grad_hess, func, grad, x0, args=(), eps=1e-4, tol=1e-4,
         maggrad = np.sum(absgrad)
         eta = min([0.5, np.sqrt(maggrad)])
         termcond = eta * maggrad
-        xsupi = np.zeros(len(x0), dtype=x0.dtype)
-        ri = fgrad
-        psupi = -ri
-        i = 0
-        dri0 = np.dot(ri, ri)
 
         # Inner loop: solve the Newton update by conjugate gradient, to
         # avoid inverting the Hessian
-        while np.sum(np.abs(ri)) > termcond:
-            Ap = fhess_p(psupi)
-            # check curvature
-            curv = np.dot(psupi, Ap)
-            if 0 <= curv <= 3*np.finfo(np.float64).eps:
+        xsupi = _cg(fhess_p, fgrad, maxiter=maxinner, tol=termcond)
+
+        alphak = 1.0
+
+        if line_search:
+            try:
+                alphak, fc, gc, old_fval, old_old_fval, gfkp1 = \
+                    _line_search_wolfe12(func, grad, xk, xsupi, fgrad,
+                                         old_fval, old_old_fval, args=args)
+            except _LineSearchError:
+                warnings.warn('Line Search failed')
                 break
-            elif curv < 0:
-                if (i > 0):
-                    break
-                else:
-                    # fall back to steepest descent direction
-                    xsupi = xsupi + dri0 / curv * psupi
-                    break
-            alphai = dri0 / curv
-            xsupi = xsupi + alphai * psupi
-            ri = ri + alphai * Ap
-            dri1 = np.dot(ri, ri)
-            betai = dri1 / dri0
-            psupi = -ri + betai * psupi
-            i = i + 1
-            dri0 = dri1          # update np.dot(ri,ri) for next time.
-
-        try:
-            alphak, fc, gc, old_fval, old_old_fval, gfkp1 = \
-                _line_search_wolfe12(func, grad, xk, xsupi, fgrad,
-                                     old_fval, old_old_fval, args=args)
-        except _LineSearchError:
-            warnings.warn('Line Search failed')
-            break
```

Comment:
`xsupi += alphai * psupi` is better than `xsupi = xsupi + alphai * psupi` in any case because it is readable, allocates less memory and is possibly faster. `psupi *= betai; psupi -= ri` is less readable, so bench it to see if it's worth it. Also be careful some arrays may be views so modifying them inplace will change the program behavior.